mm/iob/iob_free.c

```diff
@@ -151,21 +158,30 @@ FAR struct iob_s *iob_free(FAR struct iob_s *iob)
 
   spin_unlock_irqrestore(&g_iob_lock, flags);
 
-  /* Signal that an IOB is available.  If there is a thread blocked,
-   * waiting for an IOB, this will wake up exactly one thread.  The
-   * semaphore count will correctly indicated that the awakened task
-   * owns an IOB and should find it in the committed list.
+  /* Signal that an IOB is available. This is done with schedule locked
+   * to make sure that both g_iob_sem and g_throttle_sem are incremented
+   * together (if applicable). After the schedule is unlocked, if there
+   * is a thread blocked, waiting for an IOB, this will wake up exactly
+   * one thread.  The semaphore count will correctly indicate that the
+   * awakened task owns an IOB and should find it in the committed list.
    */
 
+  sched_lock();
```

Comment:
how to handle SMP case?

Reply:
`sched_lock` has its own logic to handle with the SMP case. In fact, `DEBUGASSERT(g_throttle_sem.semcount <= (CONFIG_IOB_NBUFFERS - CONFIG_IOB_THROTTLE))` was asserting when SMP was enabled (running `iperf` for a few minutes). There was a mismatch between the counters (as posting to the `g_iob_sem` semaphore may switch the context before updating the `g_throttle_sem` value). Locking the scheduler fixed the mismatch...

Reply:
A better explanation about it is in the following comment: https://github.com/apache/nuttx/pull/12501#discussion_r1641875447